### PR TITLE
Call matcher

### PR DESCRIPTION
**Pull request by Rivukis**
Something I've been working on to give swift the ability to test whether a dependency had a function called on it. Something I've grown accustomed to having while using Cedar.

After trying to use SwiftyMock with Quick and Nimble I noticed some bizarre behavior so I set out to make a much lighter version of function recording and thought it would be cool if this was already built into Nimble.

Please let me know what you think. Feel free to change the README.md changes. That was just a first stab at describing what it does and how to use it.

-Brian Radebaugh

### CODE REVIEW

**Comment by Rivukis** [comment]
I can't figure out why the build failed. Everything works locally. I don't think I have access to start another build in hopes it is just a fluke. Any help is appreciated.

-Brian Radebaugh

**Comment by jeffh** [comment]
The test failed probably because of #177, I reran the build from travis and it seems to be working. I haven't fully reviewed the PR but its API is pretty cool.

Ironically, I would be more inclined to move mocking to a separate library. I still feel It's much more sensitive to the implementation of Swift that would require more regular updates than most of Nimble's matchers. It would more likely mean moving this PR to a separate library.

All being said, I'll take a closer look this weekend. Thanks for the massive work in the PR :+1:

**Comment by Rivukis** [comment]
Thank you for the fast response. Whichever you decide (assuming you want this to be a part of Nimble) just let me know what I need to do.

I thought of making mocking easier in swift, but without the objc-runtime it's near impossible. This part of remembering called functions is assuming that whoever is going to use this already has a plan for mocking objects. I provided an example in the README.md file in case they hadn't thought of one yet.

File: Nimble/CallRecorder.swift [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    case DontCare
+    case NonNil
+    case Nil
+    case InstanceOf(type: Any.Type)
+    case InstanceOfWith(type: Any.Type, option: ArgumentOption)
+    case KindOf(type: AnyObject.Type)
+    
+    public var description: String {
+        switch self {
+        case .DontCare:
```
Comment (jeffh) [inline Nimble/CallRecorder.swift]: I'd prefer 'Anything' instead of 'DontCare'

File: NimbleTests/CallRecorderTest.swift [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+        let result1 : DidCallResult = testClass.didCall(function: "not a function", recordedCallsDescOption: .Yes)
+        let result2 : DidCallResult = testClass.didCall(function: "not a function", count: 1, recordedCallsDescOption: .Yes)
+        let result3 : DidCallResult = testClass.didCall(function: "not a function", atLeast: 1, recordedCallsDescOption: .Yes)
+        let result4 : DidCallResult = testClass.didCall(function: "not a function", atMost: -1, recordedCallsDescOption: .Yes)
+        let result5 : DidCallResult = testClass.didCall(function: "not a function", withArgs: [], recordedCallsDescOption: .Yes)
+        let result6 : DidCallResult = testClass.didCall(function: "not a function", withArgs: [], count: 1, recordedCallsDescOption: .Yes)
+        let result7 : DidCallResult = testClass.didCall(function: "not a function", withArgs: [], atLeast: 1, recordedCallsDescOption: .Yes)
+        let result8 : DidCallResult = testClass.didCall(function: "not a function", withArgs: [], atMost: -1, recordedCallsDescOption: .Yes)
+        
+        print(result1)
```
Comment (jeffh) [inline NimbleTests/CallRecorderTest.swift]: Extra print should be removed

File: NimbleTests/CallRecorderTest.swift [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
```
Comment (jeffh) [inline NimbleTests/CallRecorderTest.swift]: Extra print should be removed

File: Nimble/Matchers/Call.swift [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    if !countDescription.isEmpty {
+        let pluralism = count == 1 ? "" : "s"
+        description += " \(countDescription) \(count) time\(pluralism)"
+    }
+    
+    return description
+}
+
+private func isSuccessfulTest(didDoIt: Bool, _ isNegationTest: Bool) -> Bool {
+    return didDoIt && !isNegationTest || !didDoIt && isNegationTest
```
Comment (jeffh) [inline Nimble/Matchers/Call.swift]: Nitpick: make the ordering explicit

File: Nimble/CallRecorder.swift [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    var called: (functionList: [String], argumentsList: [[Any]]) {get set}
+    
+    // **MUST** call in every method you want to spy
+    func recordCall(function function: String, arguments: Any...)
+    
+    // Used if you want to reset the called function/arguments lists
+    func clearRecordedLists()
+    
+    
+    // For Internal Use ONLY
```
Comment (jeffh) [inline Nimble/CallRecorder.swift]: Is there a better way to make this internal without exposing these methods?

Also, make the tests use what an end user use instead of accessing internals in tests.

File: Nimble/CallRecorder.swift [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+        case .Optional:
+            return "ArgumentOption.Optional"
+        }
+    }
+}
+
+public struct DidCallResult {
+    public let success: Bool
+    public let recordedCallsDescription: String
+//    public let successfullyCalledCount: Int
```
Comment (jeffh) [inline Nimble/CallRecorder.swift]: Is this needed?

**Comment by jeffh** [comment]
I've put comments into this. Although I would be more comfortable moving this to a separate library under the quick org. I'm not really super comfortable to the string parsing of descriptions. Thoughts @quick/contributors?

Anyway. I'd recommend changing the tests to be more of how a test writer would use this API instead of using APIs marked internal. Developers will most likely look at these tests to see how to perform specific kinds of expectations.

**Comment by modocache** [comment]
A resounding "yes!" from me on making it an entirely new project under the org. Very excited to see this develop.

> Anyway. I'd recommend changing the tests to be more of how a test writer would use this API instead of using APIs marked internal. Developers will most likely look at these tests to see how to perform specific kinds of expectations.

Absolutely +1.

File: Nimble/CallRecorder.swift (hunk 1 above)
Comment (Rivukis) [inline Nimble/CallRecorder.swift]: I was going to use 'Anything' but I didn't want it to conflict with idea of swift's 'Any' or 'AnyObject'. Also I thought it would read better when using 'toNot'. If you / everyone still agrees that 'Anything' would be better then I'll change it.

File: Nimble/CallRecorder.swift (hunk 5 above)
Comment (Rivukis) [inline Nimble/CallRecorder.swift]: To be honest I'm not sure there is a better way. The mocked object needs to conform to this so that the matcher can call these functions on the mocked object. If you have an idea I'm willing to explore it, but as of now I'm not sure what I could do. That being said I'll think on it and see what/if I can come up with anything.

File: Nimble/Matchers/Call.swift (hunk 4 above)
Comment (Rivukis) [inline Nimble/Matchers/Call.swift]: Just to be clear are you talking about the fact that '&&' has precedence over '||' or are you talking about the callers of this function not knowing which to put first 'didDoIt' or 'isNegationTest'?

File: Nimble/CallRecorder.swift (hunk 6 above)
Comment (Rivukis) [inline Nimble/CallRecorder.swift]: No, sorry. This is a note to myself. It was a reminder for me to see if I wanted to include the called count of a function when the tester is wanting a specific 'count' times called (also 'atLeast' and 'atMost'). However after thinking about it I feel like it would be too much information since I already print out every called function on the object at the time it is tested (if the test fails)

**Comment by Rivukis** [comment]
About the tests. I wrote them in that way since those APIs marked as internal (like it or not because of swift) are public.

Also, if you look below those first tests you'll see that I am testing the class using the preferred APIs. Which means I could essentially get rid of those first tests and only have to rewrite the ones covering the 'clearRecordedLists' function.

I've been convinced. I'll make the change.

File: NimbleTests/CallRecorderTest.swift (hunk 2 above)
Comment (Rivukis) [inline NimbleTests/CallRecorderTest.swift]: Fixed

File: NimbleTests/CallRecorderTest.swift (hunk 3 above)
Comment (Rivukis) [inline NimbleTests/CallRecorderTest.swift]: Fixed

File: Nimble/Matchers/Call.swift (hunk 4 above)
Comment (jeffh) [inline Nimble/Matchers/Call.swift]: I'm just talking about operator precedence.

File: Nimble/CallRecorder.swift (hunk 1 above)
Comment (jeffh) [inline Nimble/CallRecorder.swift]: I still prefer `Anything` mostly because I'm not a fan of contractions in source code.

File: Nimble/CallRecorder.swift [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+    func didCall(function function: String, atMost count: Int, recordedCallsDescOption: DidCallResultIncludeOption) -> DidCallResult
+    
+    func didCall(function function: String, withArgs arguments: Array<Any>, recordedCallsDescOption: DidCallResultIncludeOption) -> DidCallResult
+    func didCall(function function: String, withArgs arguments: Array<Any>, count: Int, recordedCallsDescOption: DidCallResultIncludeOption) -> DidCallResult
+    func didCall(function function: String, withArgs arguments: Array<Any>, atLeast count: Int, recordedCallsDescOption: DidCallResultIncludeOption) -> DidCallResult
+    func didCall(function function: String, withArgs arguments: Array<Any>, atMost count: Int, recordedCallsDescOption: DidCallResultIncludeOption) -> DidCallResult
+}
+
+public extension CallRecorder {
+    func recordCall(function function: String, arguments: Any...) {
```
Comment (jeffh) [inline Nimble/CallRecorder.swift]: Is there a reason to not make `function` equal to `__FUNCTION__` by default?

**Comment by jeffh** [comment]
Hey @Rivukis,

I got a chance to play around with this more.

I seem to get a failure with differing types:

```swift
class TestClass : CallRecorder {
    var called = (functionList: [String](), argumentsList: [[Any]]())    
    func doSomething(string: String, value: Int) {
        self.recordCall(function: __FUNCTION__, arguments: [string, value])
    }
}

let testClass = TestClass()
testClass.doSomething("foo", value: 1)
expect(testClass).to(call(function: "doSomething(_:value:)", withArguments: ["foo", 1]))
```

Which produces:

```
Assertions: failed - expected to call <doSomething(_:value:)> from TestClass with foo, 1, got <doSomething(_:value:) with (foo,1)>
```

File: Nimble/CallRecorder.swift (hunk 1 above)
Comment (Rivukis) [inline Nimble/CallRecorder.swift]: fixed

File: Nimble/Matchers/Call.swift (hunk 4 above)
Comment (Rivukis) [inline Nimble/Matchers/Call.swift]: Fixed by changing the bool check to only have to make one comparison instead of one, two, or three (depending on short-circuiting). It's probably insignificant savings, but when it comes to running a lot of tests, every little bit helps.

**Comment by Rivukis** [comment]
The class that failed was in the merge conflict in the project file. I ran `git checkout --theirs Nimble.xcodeproj/project.pbxproj` and then added my files back which shouldn't have messed up anything. I also ran all the tests on all three platforms afterward with zero failures. Can you re-run the CI build?

**Comment by Rivukis** [comment]
> Hey @Rivukis,
> 
> I got a chance to play around with this more.
> 
> I seem to get a failure with differing types:
> 
> ```
> class TestClass : CallRecorder {
>     var called = (functionList: [String](), argumentsList: [[Any]]())    
>     func doSomething(string: String, value: Int) {
>         self.recordCall(function: __FUNCTION__, arguments: [string, value])
>     }
> }
> 
> let testClass = TestClass()
> testClass.doSomething("foo", value: 1)
> expect(testClass).to(call(function: "doSomething(_:value:)", withArguments: ["foo", 1]))
> ```
> Which produces:
> 
> ```
> Assertions: failed - expected to call <doSomething(_:value:)> from TestClass with foo, 1, got <doSomething(_:value:) with (foo,1)>
> ```

Hey @jeffh,
The reason you got the failure is because the ```arguments``` parameter passed to ```recordCall()``` is variadic while the ```arguments``` parameter in ```call(function:withArguments:)``` isn't. In a nutshell you passed the arguments in ```doSomething(_:value:)``` as one argument that is an array holding the two values, which means the call recorder thinks there is only one argument for that function. The call matcher is then trying to match ```["foo", 1]``` to ```"foo"``` and nothing to ```1```. _In fact there is a short-circut in the call matcher that reports back false because the number of arguments expected versus actual don't match._

I tried making the call matcher's ```arguments``` parameter also variadic but I kept running into issues. I've updated the README.md to reflect this. Also, until I can get the call matchers ```arguments``` parameter to be variadic it would probably be a good idea to not make the ```recordCall()```'s ```arguments``` variadic. This way there's less confusion.

**Comment by Rivukis** [comment]
@jeffh 
About the variadic issue you ran into earlier. I've fixed it now in the other direction. Both are now variadic. Not sure why I was having problem making them variadic before. Either way, it's solved now.

**Comment by Rivukis** [comment]
I don't think that build server likes me... Let me know if it's on my end.

**Comment by Rivukis** [comment]
I found a way to no longer depend on the CustomStringConvertible for equality between arguments being passed in and arguments being recorded. I had a better solution that would hide the "Optional : Equatable" stuff from the user but there is a bug (at least I hope it's a bug) in Swift. I've filed an issue on their Jira board: https://bugs.swift.org/browse/SR-1132

**Comment by Rivukis** [comment]
After the merge conflict I'm not able to build the project, but it's due to classes that I didn't touch. Please let me know if I need to do something on my branch.

**Comment by Rivukis** [comment]
Why does travis-ci hate me? Help please.

**Comment by modocache** [comment]
The failing jobs appear to be SwiftPM related. Here's the error from the SwiftPM jobs, on both macOS and Linux:

```
Compiling Swift Module 'Nimble' (47 sources)
/home/travis/build/Quick/Nimble/Sources/Nimble/CallRecorder.swift:290:24: error: value of type 'NSObject' has no member 'isKindOfClass'
                return recordedArgAsObject.isKindOfClass(type)
                       ^~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~
<unknown>:0: error: build had 1 command failures
error: exit(1): /home/travis/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2016-02-25-a/usr/bin/swift-build-tool -f /home/travis/build/Quick/Nimble/.build/debug.yaml default
```

I'm not sure if this version of Swift, `DEVELOPMENT-SNAPSHOT-2016-02-25-a`, defines `NSObject.isKindOfClass()`. We should probably update the Swift CI uses -- instead of a snapshot, we should use the official release of Swift 2.2.

I'm not sure why the Xcode 8 job failed, the log appears to have been cut off... are we emitting too many logs? Maybe we should condense the output using xcpretty...? @sharplet, you added Xcode 8 to the build matrix, so maybe you know what might be going on here?

**Comment by tjarratt** [comment]
Is this something that is still considered valuable? I noticed that this was in the `blocked` part of [the prioritization project](https://github.com/Quick/Nimble/projects/1)

@Rivukis would you want a second set of eyes looking at the Xcode 8 job failure on Travis? It also seems like this needs to be rebased on top of master...

I'm asking because I was starting to write a Nimble matcher to do exactly this and remembered that I had stopped a year ago because it looked like this PR was going to get merged in. Chiming in now to see how I could help out, since I'd love to be able to use this in my projects.

**Comment by jeffh** [comment]
@tjarratt, it's valuable, but there's still more work needed than the code available IIRC. I haven't looked at this in awhile. It's a lot of code that I haven't had time to fully review but, I'd be up to doing so if it was mergeable. **A smaller MVP may be easier to get merged and grow from there** (e.g. - call with only argument equality as the first step).

I will admit that I'm biased and currently prefer manual mocks. I'm willing to accept this code mostly as-is, but I personally wouldn't put much to maintain it (other @Quick/core members may have other opinions). Mostly because it doesn't feel like it fits well inside Nimble or even Swift:

 - **There's no integration to allow use of other Nimble matchers**. That's mostly my fault and I'm looking into redesigning the `Matcher` API. But I'd almost prefer just property accesses that use Nimble's normal matchers instead of overloading it all into `call`.
 - **Argument checking behavior is based on stringification**.
     - Argument equality is based on stringified values. See [line in question](https://github.com/Quick/Nimble/pull/221/files#diff-2a8c107f45c2a905116efeabc1724248R266). It just seems like a gotcha. I know it's perhaps a limitation of Swift, but it feels like surprising behavior to me.
     - Nil / Optional detection is based on parsing the description of the value? I haven't fully checked, but [it seems like it](https://github.com/Quick/Nimble/pull/221/files#diff-2a8c107f45c2a905116efeabc1724248R272).

---------------------------

Ignoring architectural complaints I have, I'm looking for at least:

 - Documentation about gotchas of the current implementation
    - [ ] Does floats work with this matcher (I'm guess probably not)?
    - [ ] A section about the stringification implementation since it's a leaky implementation detail.
    - [ ] We should probably note that this matcher has a less stable implementation than the rest of Nimble. I'm sure it'll change between Swift versions until a proper reflection API is provided.
    - [ ] This probably leaks if `self` is passed as an argument that is captured. We should note that case.
 - Significantly more test coverage of the current implementation:
    - [ ] Test `Argument.*` passed to `call` matcher
    - [ ] Test with custom class instances arguments
    - [ ] Test with custom structs as arguments
    - [ ] Test with other built-in data types than strings and ints (arrays, maps, floats, tuples)
    - [ ] Failure messages & negation tests of the above behaviors
    - [ ] Test with Objective-C classes as argument
    - [ ] Test if argument equality works with class inheritance
 - [ ] SwiftLint compliance

...for this to be mergeable.

**Comment by tjarratt** [comment]
Seems like we'd need [a redesigned Matcher API](https://github.com/Quick/Nimble/pull/390) in order to accomplish **Argument checking behavior (not) being based on stringification** -- ie: we want Nimble matchers to be able to compose other nimble matchers. The current Matcher protocol, with the associated type seems to makes it impossible to, say, have an array of `Matcher` that one would use to verify arguments to a method.

e.g.:
```swift
expect(myFake).to(haveReceived("aMethod(withAnInt:andAnotherInt:)").with(equal(5), beLessThan(8)))
```

For what it's worth, I've spent a fair amount of time writing and testing to manual mocks, and have found the assertion writing to be very tedious. It would be nice to have a matcher that could verify this behavior without 

Is this yet another reason why you opened that PR, @jeffh?

**Comment by Rivukis** [comment]
I stopped working on this because I stopped getting feed back on whether or not people wanted this to get merged in. As far as the fixes necessary to get this merged in, I'm more than happy to pick this project back up.
I think the list needs to be updated as far as what is necessary because the argument checking does NOT use stringification (I think v1 might have..?). Instead it is actually using the Swift provided protocol, `Equatable`. To get that to work when the types are unknown, I made `GloballyEquatable`. The function name is stringified but because that argument is defaulted to `#function` there isn't an issue except for function overloading.

**Comment by Rivukis** [comment]
As far as using other matchers, that would be cool, but I've never needed that functionality so I'm not sure if it is needed to get merged in but rather a feature to be added later.
As far as the manual mock, I recommend in the README a way to make this work which is to make your real class's interface a protocol and then have the real class plus the fake conform to it. The additional step is that the manual fake also conform to `CallRecorder` and call `record()` in each function. So I'm not sure what you mean when you want a manual mock that differs from the implementation.
As far as the test coverage request. Some of what you've listed is being tested. However, all equality checks are predicated on using `Equatable` so as long as the type conforms to that protocol and `GloballyEquatable` (which requires no additional work for conformance) then the any type will work. The only issue I can see is getting a tuples and closures to conform to Equatable.

**Comment by jeffh** [comment]
@Rivukis, I must of missed that through the code. I thought that's what the behavior of the program was when I last checked it out. I'll have to try again.

The matcher feature can to be deferred to a later point. The [list of requirements I mentioned earlier would make me willing to accept it](https://github.com/Quick/Nimble/pull/221#issuecomment-274705692).

I'll check out this PR and review it again tomorrow on my allocated OSS time.

**Comment by jeffh** [comment]
@tjarratt, Yes, that's part of what the Matcher API redesign is trying to address. It's worth noting that this PR will still require some work in creating your mocks. They're not as feature-rich as Cedar's mocks or OCMock. You're still required to define the methods of the mocks you create.

While I know many people would like a mock feature. I have a concern of the support issues based on expectations of mocks in other languages and what's supported.

**Comment by jeffh** [comment]
I tried running this PR, but it was a non-trivial merge beyond just naively merging the xcodeproj file to get it running on Xcode 8.2.1

**Comment by Rivukis** [comment]
Can someone help me with the build server?

**Comment by jeffh** [comment]
I'm not too familiar on the Linux side of things, but it seems like `isKind(of:)` isn't available in Linux, probably because the Objective-C runtime doesn't exist. See [beAKindOf](https://github.com/Quick/Nimble/blob/master/Sources/Nimble/Matchers/BeAKindOf.swift#L23-L48).

**Comment by Rivukis** [comment]
I decided to make a separate framework recording calls. It's call "Spry" and can be found on my GitHub page. It includes a way for test doubles to record calls and stub functions. Spry also has an optional "Spry+Nimble" cocoapod which includes a Nimble matcher for testing if a call was made. If anyone is interested, check it out and let me know what you think.

I hope I wasn't out of line, but I copied a file out of Nimble to get my Nimble matcher tests to pass (I needed the NimbleUtil.swift file so I could test the failure message.) This copied file is not a part of any cocoapod nor will it ever be. It is used in the sample project for testing purposes ONLY. I also added a comment at the top of the file so anyone who sees it knows it came from this repository. If this is not allowed, please let me know. I will remove it and find another way to test the matcher.

Because of this, I am going to close this pull request. Thank you for all the feedback along the way!

**Comment by jeffh** [comment]
Thanks for following up @Rivukis! I think it would be nice to have a wiki page on Quick that lists various supporting libraries (linked to on the README) and Spry could be there.

**Comment by Rivukis** [comment]
Can you point me in the right direction on how to get started?

Thank you,
Brian
